### ISSUE TITLE
Minor notes

### ISSUE THREAD

**Issue by jmarrec** [CambridgeEngineering/PartIA-Computing-Michaelmas#31]
I stumbled upon your notebooks as I have enrolled in an edge EDX course on data science on the built environment.

I originally didn't realize it wasn't from the course author, so I had taken some detailed notes in an attempt to be helpful. I think they could still be useful though :)

I am an experienced python coder, but I did skim through each of the notebooks (I skipped most of the exercises though). IMHO it looks like it'd indeed make a good introduction for beginners! Good stuff! 

Couple of (very) small remarks:

* 01 Variables and assignment and operator precedence.ipynb has a typo:
> "Languages have rules for what charcters can be used in variable names"

https://github.com/CambridgeEngineering/PartIA-Computing-Michaelmas/blob/1f33af59b9f92bf421260f65211b74d1977e2fa6/01%20Variables%2C%20assignment%20and%20operator%20precedence.ipynb#L628

* 02: `print(10 < 9 or 15 < 20)`. Similar to the previous operator precedence notebook I'd prefer `print((10 < 9) or (15 < 20))` to make it completely unambiguous what's happening.

* 03: I really liked the "Example: inexact representation", very powerful demonstration of roundoff errors.

* Early on, I would probably mention `help(function)` and the notebook equivalent `function?`. This is an invaluable tool to quickly know what a function does, and its arguments, without having to go check online. It would fit well in the "Library functions" notebook for eg, or even before the first time you mention that range doesn't include end. `range?` => `range(start, stop[, step]) -> range object`

* 08:
    * `RuntimeError: Unable to install quandl`. On the other hand, `!pip install quandl` shows its already fine. Might be confusing a bit. Wrap that in a `try: import quandl` like you do in 09 instead.
    * I would probably mention `fig, ax = plt.plot(...)`. I would probably mention `ax.axhline` / `ax.axvline` which is often useful to annotate a graph, include limits etc.
    * Pie charts? Really? :)

* 09, USS Yorktown Smart Ship: image broken.

* 11: complexity and timings against libraries. True, numpy probably has good heuristics to avoid worse case complexity. It's also coded in Cython checking types (https://github.com/numpy/numpy/blob/464f79eb1d05bf938d16b49da1c39a4e02506fa3/numpy/core/src/npysort/quicksort.c.src)... (`np.sort?` will also show you that it takes a 'kind' argument to use something else than quicksort, **edit**: You do use it in 12 actually). This is just a side note, I realize this is aimed at beginners.

    * Using Jupyter (or IPython) you can use 'tab-completion' to see the available attributes and methods. You will often know from the name which one you need. Perhaps that should be mentionned in one of the first notebooks (and completing existing variable names)

* 12: very minor: you mention `__repr__` but not `__str__`. There's a difference, and technically `__repr__` isn't called when print if `__str_`_ is defined... https://stackoverflow.com/questions/1436703/difference-between-str-and-repr

**Comment by jmarrec** [comment]
Feel free to close this issue whenever you want. This is just (constructive, I hope) feedback, not an "issue"

**Comment by garth-wells** [comment]
Thanks @jmarrec. The course runs again next October so it might before I fo through your helpful comments and suggestions.

I think you have been looking at an older version because I coupled of things you point out have been fixed (the Yorktown image) or removed (quandl).

**Comment by garth-wells** [comment]
@jmarrec Could you tell me which EDX course you're following?

**Comment by okkymabruri** [comment]
@garth-wells. To be honest, I also took the same course with @jmarrec on edX.
This course -> https://www.edx.org/course/Data-Science-for-Construction-Architecture-and-Engineering

I found the missing link in "03 Types, type conversions and floating point arithmetic.ipynb"

> Gangnam Style
> In 2014, Google switched from 32-bit integers to 64-bit integers to count views when the "Gangnam Style" video was viewed more than 2,147,483,647 times, which is the limit of 32-bit integers (see https://plus.google.com/+YouTube/posts/BUXfdWqu86Q).

The link has been removed by Google.
Maybe it can be replaced with https://www.wired.co.uk/article/gangnam-breaking-the-internet
